libdevice/device_imf.hpp

```diff
@@ -0,0 +1,428 @@
+    _half_internal *= rh._half_internal;
+    return *this;
+  }
+  _iml_half &operator/=(const _iml_half &rh) {
+    _half_internal /= rh._half_internal;
+    return *this;
+  }
+  _iml_half &operator++() {
+    _half_internal += 1;
+    return *this;
+  }
+  _iml_half operator++(int) {
+    _iml_half res(*this);
+    operator++();
+    return res;
+  }
+  _iml_half &operator--() {
+    _half_internal -= 1;
+    return *this;
+  }
+  _iml_half operator--(int) {
+    _iml_half res(*this);
+    operator--();
+    return res;
+  }
+
+  _iml_half operator-() {
+    _iml_half res(-_half_internal);
+    return res;
+  }
+
+  bool operator<(const _iml_half &rh) {
+    return _half_internal < rh._half_internal;
+  }
+  bool operator>(const _iml_half &rh) {
+    return _half_internal > rh._half_internal;
+  }
+#else
+  _iml_half &operator+=(const _iml_half &rh) {
+    *this = (operator float() + static_cast<float>(rh));
+    return *this;
+  }
+  _iml_half &operator-=(const _iml_half &rh) {
+    *this = (operator float() - static_cast<float>(rh));
+    return *this;
+  }
+  _iml_half &operator*=(const _iml_half &rh) {
+    *this = (operator float() * static_cast<float>(rh));
+    return *this;
+  }
+  _iml_half &operator/=(const _iml_half &rh) {
+    *this = (operator float() / static_cast<float>(rh));
+    return *this;
+  }
+  _iml_half &operator++() {
+    *this = operator float() + 1;
+    return *this;
+  }
+  _iml_half operator++(int) {
+    _iml_half res(*this);
+    operator++();
+    return res;
+  }
+  _iml_half &operator--() {
+    *this = operator float() - 1;
+    return *this;
+  }
+  _iml_half operator--(int) {
+    _iml_half res(*this);
+    operator--();
+    return res;
+  }
+
+  _iml_half operator-() {
+    _iml_half res(-operator float());
+    return res;
+  }
+
+  bool operator<(const _iml_half &rh) {
+    return operator float() < static_cast<float>(rh);
+  }
+  bool operator>(const _iml_half &rh) {
+    return operator float() > static_cast<float>(rh);
+  }
+#endif
+  _iml_half operator+(const _iml_half &rh) {
+    _iml_half res(*this);
+    res += rh;
+    return res;
+  }
+
+  _iml_half operator-(const _iml_half &rh) {
+    _iml_half res(*this);
+    res -= rh;
+    return res;
+  }
+
+  _iml_half operator*(const _iml_half &rh) {
+    _iml_half res(*this);
+    res *= rh;
+    return res;
+  }
+
+  _iml_half operator/(const _iml_half &rh) {
+    _iml_half res(*this);
+    res /= rh;
+    return res;
+  }
+  bool operator<=(const _iml_half &rh) {
+    return operator<(rh) || operator==(rh);
+  }
+  bool operator>=(const _iml_half &rh) {
+    return operator>(rh) || operator==(rh);
+  }
+
+private:
+  _iml_half_internal _half_internal;
+};
+
+static inline float __fclamp(float x, float y, float z) {
+#if defined(__LIBDEVICE_HOST_IMPL__)
+  return __builtin_fmin(__builtin_fmax(x, y), z);
+#elif defined(__SPIR__)
+  return __spirv_ocl_fclamp(x, y, z);
+#endif
+}
+
+static inline float __fma(float x, float y, float z) {
```

Comment:
Are you going to add other basic builtins like: __sqrt, __rint, __floor, __ceil, __trunc, __fabs, __fmin, __fmax, probably __rsqrt? All of them are used in IMF math implementations.